File: ruby/sample.rb
require "./LexActivator"
require "./LexStatusCodes"


# Refer to following link for LexActivator API docs:
# https://github.com/cryptlex/lexactivator-c/blob/master/examples/LexActivator.h

def init()
  
  status = LexActivator.SetProductData(LexActivator::encode_utf16("PASTE_CONTENT_OF_PRODUCT.DAT_FILE"))
  if LexStatusCodes::LA_OK != status
    puts "Error Code: #{status}"
    exit(status)
  end

  status = LexActivator.SetProductId(LexActivator::encode_utf16("PASTE_PRODUCT_ID"), LexActivator::PermissionFlags::LA_USER)
  if LexStatusCodes::LA_OK != status
    puts "Error Code: #{status}"
    exit(status)
  end

  status = LexActivator.SetReleaseVersion(LexActivator::encode_utf16("1.0.0"))
  if LexStatusCodes::LA_OK != status
    puts "Error Code: #{status}"
    exit(status)
  end
end

def activate()
  status = LexActivator.SetLicenseKey(LexActivator::encode_utf16("PASTE_LICENSE_KEY"))
  if LexStatusCodes::LA_OK != status
    puts "Error Code: #{status}"
    exit(status)
  end

  status = LexActivator.SetActivationMetadata(LexActivator::encode_utf16("key1"), LexActivator::encode_utf16("value1"))
  if LexStatusCodes::LA_OK != status
    puts "Error Code: #{status}"
    exit(status)
  end

  status = LexActivator.ActivateLicense()
  if [LexStatusCodes::LA_OK, LexStatusCodes::LA_EXPIRED, LexStatusCodes::LA_SUSPENDED].include?(status)
    puts "License activated successfully: #{status}"
  else
    puts "License activation failed: #{status}"
    exit(status)
  end
end

def activate_trial()
  status = LexActivator.SetTrialActivationMetadata(LexActivator::encode_utf16("key1"), LexActivator::encode_utf16("value1"))
  if LexStatusCodes::LA_OK != status
    puts "Error Code: #{status}"
    exit(status)
  end

  status = LexActivator.ActivateTrial()
  if LexStatusCodes::LA_OK == status
    puts "Product trial activated successfully!"
  elsif LexStatusCodes::LA_TRIAL_EXPIRED == status
    puts "Product trial has expired!"
  else
    puts "Product trial activation failed: #{status}"
  end
end

# License callback (optional)
LicenseCallback = FFI::Function.new(:void, [:uint]) do |status|
  puts "License status: #{status}"
end

SoftwareReleaseUpdateCallback = FFI::Function.new(:void, [:uint]) do |status|
  puts "Release status: #{status}"
end

# Run it
init()
LexActivator.SetLicenseCallback(LicenseCallback)
status = LexActivator.IsLicenseGenuine()
if LexStatusCodes::LA_OK == status
  # get days left for expiry
  expiryDate = FFI::MemoryPointer.new(:uint)
  LexActivator.GetLicenseExpiryDate(expiryDate)
  daysLeft = (expiryDate.read_int - Time.now.to_i) / 86400
  puts "Days left: #{daysLeft}"
  tier = FFI::MemoryPointer.new(:int64)
  LexActivator.GetLicenseEntitlementSetTier(tier)
  puts "tier: #{tier.read_int64}"
  # get license user email
  buffer = FFI::MemoryPointer.new(:char, 256)
  LexActivator.GetLicenseUserEmail(buffer, buffer.size)
  email = LexActivator::decode_utf16(buffer.read_string(buffer.size).rstrip)
  puts "License user email: #{email}"
  puts "License is genuinely activated!"
elsif LexStatusCodes::LA_EXPIRED == status
  puts "License is genuinely activated but has expired!"
elsif LexStatusCodes::LA_SUSPENDED == status
  puts "License is genuinely activated but has been suspended!"
elsif LexStatusCodes::LA_GRACE_PERIOD_OVER == status
  puts "License is genuinely activated but grace period is over!"
else
  trialStatus = LexActivator.IsTrialGenuine()
  if LexStatusCodes::LA_OK == trialStatus
    # get days left for expiry
    trialExpiryDate = FFI::MemoryPointer.new(:uint)
    LexActivator.GetTrialExpiryDate(trialExpiryDate)
    daysLeft = (trialExpiryDate.read_int - Time.now.to_i) / 86400
    puts "Trial days left: #{daysLeft}"
  elsif LexStatusCodes::LA_TRIAL_EXPIRED == trialStatus
    puts "Trial has expired!"
    # Time to buy the license and activate the app
    activate()
  else
    puts "Either trial has not started or has been tampered!"
    # Activating the trial
    activate_trial()
  end
end

# Checking for software release update
# Call SetReleaseVersion(), SetReleasePlatform() and SetReleaseChannel() before calling CheckReleaseUpdate()
# Release version, platform and channel must be set before checking for an update
# status = LexActivator.SetReleasePlatform(LexActivator::encode_utf16("windows"))
# status = LexActivator.SetReleaseChannel(LexActivator::encode_utf16("1.0.0"))
# status = LexActivator.CheckForReleaseUpdate("windows", "1.0.0", "stable", SoftwareReleaseUpdateCallback)

puts "Press Enter to exit..."
gets

File: go/sample.go
package main

import (
	"bufio"
	"fmt"
	"os"

	"github.com/cryptlex/lexactivator-go"
)

// server sync license callback
func licenseCallback(status int) {
	if status == lexactivator.LA_OK {
		fmt.Println("License is genuinely activated.")
	} else if status == lexactivator.LA_EXPIRED {
		fmt.Println("License is genuinely activated, but has expired")
	} else if status == lexactivator.LA_SUSPENDED {
		fmt.Println("License is genuinely activated, but has been suspended")
	} else {
		fmt.Println("License error status:", status)
	}
}

func softwareReleaseUpdateCallback(status int, release *lexactivator.Release, userData interface{}) {
	if status == lexactivator.LA_RELEASE_UPDATE_AVAILABLE {
		fmt.Println("A new update is available for the app!")
		fmt.Println("Release notes: ", release.Notes)
	} else if status == lexactivator.LA_RELEASE_UPDATE_AVAILABLE_NOT_ALLOWED {
		fmt.Println("A new update is available for the app but it's not allowed!")
		fmt.Println("Release notes: ", release.Notes)
	} else if status == lexactivator.LA_RELEASE_UPDATE_NOT_AVAILABLE {
		fmt.Println("Current version is already latest!")
	} else {
		fmt.Println("Error code: ", status)
	}
}

func initData() {
	var status int
	status = lexactivator.SetProductData("PASTE_CONTENT_OF_PRODUCT.DAT_FILE")
	if lexactivator.LA_OK != status {
		fmt.Println("Error Code:", status)
		os.Exit(1)
	}

	status = lexactivator.SetProductId("PASTE_PRODUCT_ID", lexactivator.LA_USER)
	if lexactivator.LA_OK != status {
		fmt.Println("Error Code:", status)
		os.Exit(1)
	}
	// Set this to the release version of your app
	status = lexactivator.SetReleaseVersion("1.0.0")
	if lexactivator.LA_OK != status {
		fmt.Println("Error Code:", status)
		os.Exit(1)
	}

}

func activate() {
	var status int
	status = lexactivator.SetLicenseKey("PASTE_LICENSE_KEY")
	if lexactivator.LA_OK != status {
		fmt.Println("Error Code:", status)
		os.Exit(1)
	}

	status = lexactivator.SetActivationMetadata("key1", "value1")
	if lexactivator.LA_OK != status {
		fmt.Println("Error Code:", status)
		os.Exit(1)
	}

	status = lexactivator.ActivateLicense()
	if lexactivator.LA_OK == status || lexactivator.LA_EXPIRED == status || lexactivator.LA_SUSPENDED == status {
		fmt.Println("License activated successfully:", status)
	} else {
		fmt.Println("License activation failed:", status)
	}
}

func activateTrial() {
	var status int
	status = lexactivator.SetTrialActivationMetadata("key1", "value1")
	if lexactivator.LA_OK != status {
		fmt.Println("Error Code:", status)
		os.Exit(1)
	}

	status = lexactivator.ActivateTrial()
	if lexactivator.LA_OK == status {
		fmt.Println("Product trial activated successfully!")
	} else if lexactivator.LA_TRIAL_EXPIRED == status {
		fmt.Println("Product trial has expired!")
	} else {
		fmt.Println("Product trial activation failed:", status)
	}
}

func main() {
	initData()
	var status int
	lexactivator.SetLicenseCallback(licenseCallback)
	status = lexactivator.IsLicenseGenuine()
	if lexactivator.LA_OK == status {
		var expiryDate uint
		lexactivator.GetLicenseExpiryDate(&expiryDate)
		fmt.Println("License expiry timestamp:", expiryDate)
		fmt.Println("License is genuinely activated!")
	} else if lexactivator.LA_EXPIRED == status {
		fmt.Println("License is genuinely activated but has expired!")
	} else if lexactivator.LA_SUSPENDED == status {
		fmt.Println("License is genuinely activated but has been suspended!")
	} else if lexactivator.LA_GRACE_PERIOD_OVER == status {
		fmt.Println("License is genuinely activated but grace period is over!")
	} else {
		var trialStatus int
		trialStatus = lexactivator.IsTrialGenuine()
		if lexactivator.LA_OK == trialStatus {
			var trialExpiryDate uint
			lexactivator.GetTrialExpiryDate(&trialExpiryDate)
			fmt.Println("Trial expiry timestamp:", trialExpiryDate)
		} else if lexactivator.LA_TRIAL_EXPIRED == trialStatus {
			fmt.Println("Trial has expired!")

			// Time to buy the license and activate the app
			activate()
		} else {
			fmt.Println("Either trial has not started or has been tampered!")
			// Activating the trial
			activateTrial()
		}
	}
	// Checking for software release update
	// Call SetReleaseVersion(), SetReleasePlatform() and SetReleaseChannel() before calling CheckReleaseUpdate()
	// Release version, platform and channel must be set before checking for an update
	// if lexactivator.LA_OK != status {
	// 	fmt.Println("Error Code:", status)
	// 	os.Exit(1)
	// }
	// status = lexactivator.SetReleaseChannel("RELEASE_CHANNEL")
	// if lexactivator.LA_OK != status {
	// 	fmt.Println("Error Code:", status)
	// 	os.Exit(1)
	// }
	// status = lexactivator.CheckReleaseUpdate(softwareReleaseUpdateCallback, lexactivator.LA_RELEASES_ALL, nil);
	// if lexactivator.LA_OK != status {
	// 	fmt.Println("Error Code:", status)
	// 	os.Exit(1)
	// }

	fmt.Println("Press Enter to exit...")
	bufio.NewReader(os.Stdin).ReadString('\n')
}
